File: frontend/src/api/localImport.ts
// frontend/src/api/localImport.ts
import client from './client';
import type {
  RemoteFileEntry,
  ServerBrowseResponse,
  ImportFileProgress,
  ServerImportProgress,
} from './serverImport';

// Re-export shared types for convenience
export type { RemoteFileEntry, ServerBrowseResponse, ImportFileProgress, ServerImportProgress };

// --- Local import types ---

export interface LocalBrowseRequest {
  path: string;
}

export interface LocalImportRequest {
  filePaths: string[];
  useSymlink?: boolean;
}

export interface LocalImportStartedResponse {
  importId: string;
  fileCount: number;
  message: string;
}

// --- API Functions ---

export async function browseLocal(
  experimentId: number,
  req: LocalBrowseRequest,
): Promise<ServerBrowseResponse> {
  const { data } = await client.post<ServerBrowseResponse>(
    `/experiments/${experimentId}/local-import/browse`,
    req,
  );
  return data;
}

export async function startLocalImport(
  experimentId: number,
  req: LocalImportRequest,
): Promise<LocalImportStartedResponse> {
  const { data } = await client.post<LocalImportStartedResponse>(
    `/experiments/${experimentId}/local-import/start`,
    req,
  );
  return data;
}

export async function getLocalImportProgress(
  experimentId: number,
  importId: string,
): Promise<ServerImportProgress> {
  const { data } = await client.get<ServerImportProgress>(
    `/experiments/${experimentId}/local-import/${importId}/progress`,
  );
  return data;
}

File: backend/schemas/local_import.py
# backend/schemas/local_import.py
from pydantic import Field

from schemas.common import CamelModel


class LocalBrowseRequest(CamelModel):
    """Browse a directory on the server's local filesystem."""

    path: str = "/data"


class LocalImportRequest(CamelModel):
    """Import FASTQ files from local filesystem paths."""

    file_paths: list[str] = Field(min_length=1)
    use_symlink: bool = False


class LocalImportStartedResponse(CamelModel):
    """Returned after starting a background local import."""

    import_id: str
    file_count: int
    message: str
